### ISSUE TITLE
The mirror example doesn't support zip file

### ISSUE THREAD

**Issue by shi-yan** [dokan-dev/dokany#106]
The mirror example doesn't support zip file. 

after running the example, you can select few files and right click, choose send to compressed file. you will see an error saying file_name is an empty directory. it thinks the files are empty folders somehow.

**Comment by Liryna** [comment]
Can you provide logs from mirror ?
https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan#mirrorfs

**Comment by shi-yan** [comment]
I will, but I will need sometime.

btw, the mirror example in dokan.net doesn't seem to have this issue.

**Comment by Liryna** [comment]
Oh ho, this is strange. Normally mirror c support more features.

**Comment by shi-yan** [comment]
so I compared the c# implementation and the c++ implementation.

one difference I noticed is that when creating a directory, the c# version calls into CreateDirectory() function, whereas the C++ program calls into CreateFile();

**Comment by shi-yan** [comment]
can I ask you a related question, I wrote a simple mirror filesystem. but instead of using windows api, I'm using Qt.

I have an issue, I seem to return the correct file size in GetFileInformation and FindFiles. and I can see the file size correctly from the file browser. but if I check properties, the file size is zero, and there is no creation date.

I can't figure out what's wrong.

as shown in image:

![untitled](https://cloud.githubusercontent.com/assets/326807/11109133/36a9b0c6-88a2-11e5-9195-b35fd5374f55.png)


however if you check the detail tab of the same file, you will see the file size.

![untitled2](https://cloud.githubusercontent.com/assets/326807/11109139/48d17d7e-88a2-11e5-891d-ed7858c627fa.png)

**Comment by Liryna** [comment]
Hum hard to say but I guess one of the call is probably failing or the information is not correctly provided. Have you seen this behaviour in the mirror ?

You should enable the dokan driver logs to help you during your dev:
https://github.com/dokan-dev/dokany/wiki/How-to-Debug-Dokan#dokan-driver

**Comment by suy** [comment]
My filesystem is implemented using Qt (although that's not relevant per se, since the callback implementations are just C), and I can confirm that I get correct behaviour. I had to create or find online some helper functions to convert from Qt types to Windows native ones where those are not available, but that should be irrelevant for the size or the attributes.

**Comment by shi-yan** [comment]
I think the issue I have is because I somehow makes windows think the file is a folder.

folder doesn't have "size" under that tab. but file has.

if I try to create a zip file out of the file, I get the error directory m:\ddddd.txt is empty. but ddddd.txt is not a directory.

**Comment by Liryna** [comment]
I made the test of the ZIP file with the mirror C and it work on my side (direct opening with explorer and extract all).

There should be something that your code make different from the mirror. Add some logs on your code and be sure about the information that you provide to the driver.
Dokan only forward the information you gave him to the kernel.

**Comment by shi-yan** [comment]
sorry for the red herring, I resynced the official code, it does seem to work.

but my own code still can't .

one issue I identified is that FindFirstFile fails with my implementation. from the dokany console, all I can see is this

###Create 0016
CreateFile status = 3221225626

it seems that it tries to create a file, but it doesn't call into my MirrorCreateFile at all, but already gets a return code 3221225626 which is file not found.

:(

the Mirror example is nice, but it's a bit cheating as it uses windows api to implement windows api. I hope there could be an example based on Qt or boost fs. 

I need to implement a remote filesystem. can't use windows api on linux and mac.

**Comment by suy** [comment]
El Wednesday 18 November 2015, Shi Yan escribió:
> the Mirror example is nice, but it's a bit cheating as it uses windows api
> to implement windows api. I hope there could be an example based on Qt or
> boost fs. 
> 
> I need to implement a remote filesystem. can't use windows api on linux and
> mac.

You'll have to return values to Dokan in the shape of the Windows API anyway. 
And if your project is cross-platform, you'll have to use the UNIX API on OS X 
or Linux. The examples for FUSE are pretty much the same: a file system that 
maps values from a different part of the file system.

I suggest you to subscribe to the Dokan mailing list/group and questions there 
if you have doubts. I'll try to answer. Otherwise I can try to write a wiki 
page with some words, but I'm too busy so I can't promise that.

**Comment by Liryna** [comment]
@shi-yan , @suy is right. Could you use the mailing list/group for getting help in your implementation ?
If you don't want to use the Windows API, there is the dokan FUSE https://github.com/dokan-dev/dokany/wiki/FUSE

```
###Create 0016
CreateFile status = 3221225626 == 0xC000009A //STATUS_INSUFFICIENT_RESOURCES
```

About Qt and boost, I am sure there is enough exemple how to convert Win32 define to their API on internet.

**Comment by shi-yan** [comment]
Hi guys, 

thank you for the reply.

when I solve my issues, I will release my code as an example.


I made some progress after debugging dokan.dll. the reason I saw the insufficient_resources error is because the example I'm following doesn't provide a "opendirectory" implementation.

so it failed here:
'''
		else if(disposition == FILE_OPEN) 
		{
			if (DokanInstance->DokanOperations->OpenDirectory) 
			{
                                // missing opendirectory implementation
				status = DokanInstance->DokanOperations->OpenDirectory(
					EventContext->Operation.Create.FileName, &fileInfo);
			}
		} 
'''
as soon as I implement that function, I could now create zip and use findfirstfile.

I double checked the official mirror.c in here

https://github.com/dokan-dev/dokany/blob/master/dokan_mirror/mirror.c

it indeed doesn't provided "opendirectory" function. but it works. I did see it calls createdirectory, but not directly called by dokan. it is called by createfile.

why does the example mirror.c calls into createfile, whereas my code calls into opendirectory, I still don't understand.

I know the linux fuse filesystem. what I want to achieve is mirroring a linux folder or a mac folder to windows through network. so on the windows side, I still need dokan. but on linux, I will use other fs libraries.

yes, I will subscribe to the mailing list. Thanks!
